solr/core/src/java/org/apache/solr/response/transform/EqualTermsDocTransformer.java

```diff
@@ -0,0 +1,152 @@
+  private final SchemaField sourceField;
+  private final String compareValue;
+  private final List<String> compareValueTerms;
+
+  public EqualTermsDocTransformer(String name, SchemaField sourceField, String compareValue) {
+    this.name = name;
+    this.sourceField = sourceField;
+    this.compareValue = compareValue;
+
+    // Analyze the comparison value up front
+    FieldType fieldType = sourceField.getType();
+    Analyzer analyzer = fieldType.getIndexAnalyzer();
+
+    this.compareValueTerms = analyzeToTerms(analyzer, compareValue);
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public boolean needsSolrIndexSearcher() {
+    return true;
+  }
+
+  @Override
+  public String[] getExtraRequestFields() {
+    return new String[] {sourceField.getName()};
+  }
+
+  @Override
+  public void transform(SolrDocument doc, int docid) throws IOException {
+    String fieldValue = getFieldValue(doc);
+    if (fieldValue == null) {
+      doc.setField(name, false);
+      return;
+    }
+
+    FieldType fieldType = sourceField.getType();
+    Analyzer analyzer = fieldType.getIndexAnalyzer();
+
+    // Compare terms on-the-fly as we iterate through the token stream
+    // This allows early termination as soon as we find a mismatch
+    boolean isEqual = compareTokensOnTheFly(analyzer, fieldValue);
+    doc.setField(name, isEqual);
+  }
+
+  /**
+   * Gets the string field value, or null if not present or if found a list of values other than 1.
+   */
+  private String getFieldValue(SolrDocument doc) {
+    Object fieldValue = doc.getFieldValue(sourceField.getName());
+    if (fieldValue instanceof List<?> list) {
+      if (list.size() != 1) return null;
+      fieldValue = list.getFirst();
+    }
+    if (fieldValue instanceof CharSequence) return fieldValue.toString();
+    if (fieldValue instanceof StoredField storedField) {
+      return storedField.stringValue();
+    }
+    return null;
+  }
+
+  /**
+   * Compares tokens from the analyzed text with the pre-analyzed comparison tokens. Returns false
+   * as soon as a mismatch is found.
+   */
+  private boolean compareTokensOnTheFly(Analyzer analyzer, String text) throws IOException {
+    Iterator<String> compareIter = compareValueTerms.iterator();
+
+    try (TokenStream tokenStream = analyzer.tokenStream(sourceField.getName(), text)) {
+      CharTermAttribute termAttr = tokenStream.addAttribute(CharTermAttribute.class);
+      tokenStream.reset();
+
+      while (tokenStream.incrementToken()) {
+        // Check if we've seen more tokens than in our comparison list
+        if (!compareIter.hasNext()) {
+          return false; // More tokens in source than in comparison value
+        }
+
+        // Compare the current token with the corresponding token in our pre-analyzed list
+        String currentToken = termAttr.toString();
+        String compareToken = compareIter.next();
+        if (!currentToken.equals(compareToken)) {
```

Comment:
could be improved to avoid producing a String... maybe there's a utility method to do such a comparison concisely